File: request-bodies/gps-tracks-csv-upload/scripts/convert_json_params_to_curl.py
import json
import sys
import os
import urllib.parse

def convert_value(value):
    if isinstance(value, bool):
        return str(value).lower()
    return value

def encode_to_curl(input_file, output_file, base_url):
    with open(input_file, 'r', encoding='utf-8') as file:
        json_data = json.load(file)

    params = '&'.join(
        [f"{urllib.parse.quote(key)}={urllib.parse.quote(str(convert_value(value)))}" for key, value in json_data.items()]
    )

    headers = {
        'Content-Type': 'text/csv',
        'x-api-key': '{{apiKey}}',
    }

    headers_str = ' '.join([f"-H '{key}: {value}'" for key, value in headers.items()])

    curl_command = f"curl -X POST {base_url}?{params} {headers_str} --data-binary '@'"

    with open(output_file, 'w', encoding='utf-8') as output:
        output.write(curl_command)

    print(f"Generated cURL command for {input_file}: {output_file}")

# Process command line arguments
if len(sys.argv) < 3:
    print(f"Usage: python {sys.argv[0]} <input-file> <output-file>")
    sys.exit(1)

input_file = sys.argv[1]
output_file = sys.argv[2]
base_url = 'https://apis.tollguru.com/toll/v2/gps-tracks-csv-upload'

encode_to_curl(input_file, output_file, base_url)

print('Conversion complete.')

File: request-bodies/03-TollTally-GPS-Tracks-To-Toll/scripts/convert_json_params_to_curl.js
const fs = require('fs');
const path = require('path');

function convertValue(value) {
    if (typeof value === 'boolean') {
        return String(value).toLowerCase();
    }
    return value;
}

function encodeToCurl(inputFile, outputFile, baseUrl) {
    const jsonData = JSON.parse(fs.readFileSync(inputFile, 'utf8'));

    const params = Object.entries(jsonData)
        .map(([key, value]) => {
            const val = convertValue(value);
            return `${encodeURIComponent(key)}=${encodeURIComponent(
                typeof val === 'object' ? JSON.stringify(val) : val
            )}`;
        })
        .join('&');

    // Add headers here (modify as needed)
    const headers = {
        'Content-Type': 'text/csv',
        'x-api-key': '{{apiKey}}',
    };

    const headersStr = Object.entries(headers)
        .map(([key, value]) => `-H '${key}: ${value}'`)
        .join(' ');

    const curlCommand = `curl -X POST ${baseUrl}?${params} ${headersStr} --data-binary '@'`;

    fs.writeFileSync(outputFile, curlCommand);

    console.log(`Generated cURL command for ${inputFile}: ${outputFile}`);
}

// Process command line arguments
const program = path.basename(process.argv.at(1));
const args = process.argv.slice(2);
if (args.length < 2) {
    console.error(`Usage: node ${program} <input-file> <output-file>`);
    process.exit(1);
}

const inputFile = args[0];
const outputFile = args[1];
const baseUrl = 'https://apis.tollguru.com/toll/v2/gps-tracks-csv-upload';

encodeToCurl(inputFile, outputFile, baseUrl);

console.log('Conversion complete.');

